libcache/test_libcache.c

```diff
@@ -0,0 +1,989 @@
+	cachectx_t *cache = NULL;
+
+	cache = cache_init(LIBCACHE_SRC_MEM_SIZE, size, lineSize, test_writeCb, test_readCb);
+	TEST_ASSERT_NULL(cache);
+}
+
+/* false, false -> false (no short-circuit) */
+/* covers both ifs */
+TEST(test_init, cache_init_sizes_not_zero_bytes)
+{
+	int ret = -1;
+	size_t size = LIBCACHE_CACHE_SIZE, lineSize = LIBCACHE_LINE_SIZE;
+	cachectx_t *cache = NULL;
+
+	cache = cache_init(LIBCACHE_SRC_MEM_SIZE, size, lineSize, test_writeCb, test_readCb);
+	TEST_ASSERT_NOT_NULL(cache);
+
+	ret = cache_deinit(cache);
+	TEST_ASSERT_EQUAL_INT(EOK, ret);
+}
+
+TEST_GROUP_RUNNER(test_init)
+{
+	RUN_TEST_CASE(test_init, cache_init_sizes_zero_bytes);
+	RUN_TEST_CASE(test_init, cache_init_line_size_zero_bytes);
+	RUN_TEST_CASE(test_init, cache_init_size_mod_line_size_not_zero);
+	RUN_TEST_CASE(test_init, cache_init_size_mod_line_size_zero);
+	RUN_TEST_CASE(test_init, cache_init_sizes_not_zero_bytes);
+}
+
+TEST_GROUP(test_deinit);
+
+TEST_SETUP(test_deinit)
+{
+}
+
+TEST_TEAR_DOWN(test_deinit)
+{
+}
+
+TEST(test_deinit, cache_deinit_uninitialized_cache)
+{
+	int ret;
+	cachectx_t *cache = NULL;
+
+	ret = cache_deinit(cache);
+	TEST_ASSERT_EQUAL_INT(-EINVAL, ret);
+}
+
+TEST(test_deinit, cache_deinit_initalized_cache)
+{
+	int ret;
+	cachectx_t *cache;
+
+	cache = cache_init(LIBCACHE_SRC_MEM_SIZE, LIBCACHE_CACHE_SIZE, LIBCACHE_LINE_SIZE, test_writeCb, test_readCb);
+	TEST_ASSERT_NOT_NULL(cache);
+
+	ret = cache_deinit(cache);
+	TEST_ASSERT_EQUAL_INT(EOK, ret);
+}
+
+TEST_GROUP_RUNNER(test_deinit)
+{
+	RUN_TEST_CASE(test_deinit, cache_deinit_uninitialized_cache);
+	RUN_TEST_CASE(test_deinit, cache_deinit_initalized_cache);
+}
+
+TEST_GROUP(test_read_write);
+
+TEST_SETUP(test_read_write)
+{
+	int ret;
+
+	fileRead = open("/var/read.txt", O_RDONLY | O_CREAT);
+	TEST_ASSERT_NOT_EQUAL_INT(-1, fileRead);
+
+	fileWrite = open("/var/write.txt", O_RDWR | O_CREAT);
+	TEST_ASSERT_NOT_EQUAL_INT(-1, fileWrite);
+
+	offBitsNum = LOG2(LIBCACHE_LINE_SIZE);
+	offMask = ((uint64_t)1 << offBitsNum) - 1;
+}
+
+TEST_TEAR_DOWN(test_read_write)
+{
+	int ret = -1;
+
+	ret = close(fileRead);
+	TEST_ASSERT_EQUAL_INT(0, ret);
+
+	ret = close(fileWrite);
+	TEST_ASSERT_EQUAL(0, ret);
+
+	ret = remove("/var/write.txt");
+	TEST_ASSERT_NOT_EQUAL_INT(-1, ret);
+}
+
+/*
+	if (cache == NULL || buffer == NULL || (policy != LIBCACHE_WRITE_BACK && policy != LIBCACHE_WRITE_THROUGH)) {
+		return -1;
+	}
+*/
+
+/* cases:	true, true, true -> true (short-circuit)
+			true, true, false -> true (short-circuit)
+			true, false, false -> true (short-circuit)
+			false, true, true -> true (no short-circuit)
+			false, false, true -> true (no short-circuit)
+			false, false, false -> false (no short-circuit)
+ */
+
+/* true, true, true -> true (short-circuit) */
+TEST(test_read_write, cache_write_uninitialized_cache)
+{
+	ssize_t written = 0;
+	cachectx_t *cache = NULL;
+	size_t count = 10 * sizeof(char);
+	uint64_t addr = 1143552ULL;
```

Comment:
This is a random address with no particular meaning. It reoccurs in a few places due to the fact that similar test bodies are partial copies of each other. If necessary, I can change them to avoid confusion in the future.

Please advise.

Reply:
I've grouped addresses with the same values and created suitable defines. Thank you.